File: decorator/c++/whip-decorator.cpp
#include <iostream>
#include "whip-decorator.h"

WhipDecorator::WhipDecorator(Coffee* decoratedCoffee) : CoffeeDecorator(decoratedCoffee) {};

double WhipDecorator::getCost() {
    return CoffeeDecorator::getCost() + 2;
}

std::string WhipDecorator::getIngredients() {
    return CoffeeDecorator::getIngredients() + " Whip";
}

File: decorator/js/whip-decorator.js
exports.decorator = function(coffee) {
    var originalCost = coffee.getCost;
    var originalIngredients = coffee.getIngredients;

    coffee.getCost = function() {
        return originalCost.apply(this, arguments) + 2;
    }
    coffee.getIngredients = function() {
        return originalIngredients.apply(this, arguments) + " Whip";
    }

    return coffee;
}